pyfr/solvers/navstokes/elements.py

```diff
@@ -8,6 +8,26 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
     # Use the density field for shock sensing
     shockvar = 'rho'
 
+    @staticmethod
+    def grad_con_to_pri(cons, grad_cons, cfg):
+        rho = cons[0]
+        grad_rho = grad_cons[0]
+        grad_E = grad_cons[-1]
+        # Divide momentum components by rho
+        vs = [rhov/rho for rhov in cons[1:-1]]
+
+        # Velocity gradients
+        grad_v = [(grad_rhov - v*grad_rho)/rho for grad_rhov, v in zip(grad_cons[1:-1], vs)]
+
+        # Pressure gradient
+        gamma = cfg.getfloat('constants', 'gamma')
+        grad_p = grad_E - 0.5*sum(v*v for v in vs)*grad_rho
+        for vel, grad_vel in zip(vs, grad_v):
```

Comment:
Can avoid multiply through by rho by using conservative grads

Reply:
I think np.dot here with the right set of arguments could be used here (so it performs a bunch of smaller dot products for us).